### PR TITLE
Clean up KeyValuePairUtil

### PR DESCRIPTION
**Pull request by tmat**
Removes the need to add conditional directive  

```
#if !NETCOREAPP
using Roslyn.Utilities
#endif
```

in code that uses KeyValuePair deconstructor.

`KeyValuePair` is a type in `System.Collections.Generic` in `netcoreapp3.1`, but is not available in `netstandard2.0` or `net472`. Define the type in the same namespace if it's not defined already.

`KeyValuePair<K,V>.Deconstruct` is also available in `netcoreapp3.1`, but is not available in `netstandard2.0` or `net472`. Define it as an extension method in global namespace (to be available in all our code without TFM-conditional using) if it's not defined already.

The core of the change is in [KeyValuePairUtil.cs](https://github.com/dotnet/roslyn/pull/43579/files#diff-ee3dd515898db6c107adb85ae5e78f7e), the rest is mechanical.

### CODE REVIEW

**Comment by tmat** [comment]
@mavasani

File: src/Compilers/Core/Portable/InternalUtilities/KeyValuePairUtil.cs [hunk 1]
```diff
@@ -6,21 +6,33 @@
...
+            => KeyValuePair.Create(tuple.Item1, tuple.Item2);
+    }
+}
 
-        public static void Deconstruct<TKey, TValue>(this KeyValuePair<TKey, TValue> keyValuePair, out TKey key, out TValue value)
-        {
-            key = keyValuePair.Key;
-            value = keyValuePair.Value;
-        }
+#if NETFRAMEWORK || NETSTANDARD2_0
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/KeyValuePairUtil.cs:18]:
```suggestion
#if !NETCOREAPP
```

**Comment by tmat** [comment]
Sure, I think this change actually addresses all of the objections listed there.

**Comment by tmat** [comment]
Hmm the bootstrapped .NET Core build is failing because `artifacts\Bootstrap\tasks\netcoreapp3.1\bincore\Microsoft.CodeAnalysis.VisualBasic.dll` is `netstandard2.0` version, not `netcoreapp3.1`. Looks like we are not multi-targeting VB compiler. Hmm, did not know that. This is isn't going to work w/o multi-targeting.

**Comment by tmat** [comment]
Interesting, targeting VB compiler to `netcoreapp3.1` actually does not work without further changes:

```
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Analysis\FlowAnalysis\DataFlowPass.vb(431,34): error BC30455: Argument not specified for parameter 'FileNumber' of 'Public Overloads Function Loc(FileNumber As Integer) As Long'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Analysis\FlowAnalysis\DataFlowPass.vb(432,59): error BC30455: Argument not specified for parameter 'FileNumber' of 'Public Overloads Function Loc(FileNumber As Integer) As Long'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Compilation\VisualBasicCompilation.vb(2031,34): error BC30068: Expression is a value and therefore cannot be the target of an assignment. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Compilation\VisualBasicCompilation.vb(2032,41): error BC30456: 'WithLocation' is not a member of 'ErrObject'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\CommandLine\VisualBasicCommandLineParser.vb(1802,58): error BC30521: Overload resolution failed because no accessible 'TrimEnd' is most specific for these arguments: [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Emit\PEModuleBuilder.vb(300,22): error BC30455: Argument not specified for parameter 'FileNumber' of 'Public Overloads Function Loc(FileNumber As Integer) As Long'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Emit\PEModuleBuilder.vb(301,20): error BC30455: Argument not specified for parameter 'FileNumber' of 'Public Overloads Function Loc(FileNumber As Integer) As Long'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Emit\PEModuleBuilder.vb(301,112): error BC30455: Argument not specified for parameter 'FileNumber' of 'Public Overloads Function Loc(FileNumber As Integer) As Long'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Emit\PEModuleBuilder.vb(302,30): error BC30455: Argument not specified for parameter 'FileNumber' of 'Public Overloads Function Loc(FileNumber As Integer) As Long'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Compilation\DocumentationComments\DocumentationCommentCompiler.Includes.vb(634,135): error BC30521: Overload resolution failed because no accessible 'TrimEnd' is most specific for these arguments: [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Compilation\DocumentationComments\DocumentationCommentWalker.vb(297,87): error BC30521: Overload resolution failed because no accessible 'TrimEnd' is most specific for these arguments: [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Syntax\SyntaxFactory.vb(1081,22): error BC30068: Expression is a value and therefore cannot be the target of an assignment. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Syntax\SyntaxFactory.vb(1082,40): error BC30456: 'Code' is not a member of 'ErrObject'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Syntax\SyntaxFactory.vb(1113,22): error BC30068: Expression is a value and therefore cannot be the target of an assignment. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Syntax\SyntaxFactory.vb(1114,40): error BC30456: 'Code' is not a member of 'ErrObject'. [F:\workspace\_work\1\s\src\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj]
```

**Comment by jaredpar** [comment]
@tmat 

> Looks like we are not multi-targeting VB compiler. Did not know that. 

It's due to a bug in the SDK when you build a VB add against `netcoreapp3.1`. Until that is resolved I can't see a good way of doing this. I tried some MSBuild magic but couldn't find a good solution.

https://github.com/dotnet/sdk/issues/10591

**Comment by jaredpar** [comment]
@sharwell 

> Make sure this gets approval from @jaredpar because I requested this change in #28037 (comment) and it was rejected.

@jaredpar from two years ago was definitely against this change and reading his logic I get where he was coming from

@jaredpar from today is more okay with this though. 

Sorry if this seems hypocritical but I think that a couple of factors have changed the way I feel about this:
1. We're multi-targeting now and that gives us more flexibility with these types of solutions
1. We really crossed the boundary of adding types in the `System` namespace once we started embracing `IAsyncEnumerable`. Those were a functional necessity where as `KeyValuePair` was more a convenience issue. That though pretty much ended one of my objections particularly since this is `internal` hence we're not exporting it anywhere.

**Comment by sharwell** [comment]
@tmat it won't let me dismiss the change request while it's still in draft. I have no objection to going back to the old name.

**Comment by tmat** [comment]
@jaredpar No worries, things change. 

Re multi-targeting VB: Did adding a target to the compiler's vbproj that filters out the assembly reference we don't want not work?

**Comment by jaredpar** [comment]
@tmat 

> Re multi-targeting VB: Did adding a target to the compiler's vbproj that filters out the assembly reference we don't want not work?

I worked with Nick at the time to filter this out but we couldn't find a satisfactory way to do so. We also didn't spend more than a few hours on this because it wasn't a blocking issue for the more immediate multi-targeting goal. It's possible there is a way to do this that we just missed.

**Pull request closed without merging**